aaelf64/aaelf64.rst

```diff
@@ -1249,10 +1249,57 @@ Finally the bit-field of X specified in the table (those bits of X picked out by
 
 For "``MOVW``" type relocations it is the assembler’s responsibility to encode the hw bits (bits 21 and 22) to indicate the bits in the target value that the immediate field represents.
 
-Relocation relaxation
-^^^^^^^^^^^^^^^^^^^^^
+Relocation optimization
+^^^^^^^^^^^^^^^^^^^^^^^
+
+Linkers may optionally optimize instructions affected by relocation. Relocation optimizations improve the efficiency of relocated instructions without changing their visible behaviour. There are several classes of relocation optimizations:
+
+- A single relocation optimization may change an instruction after relocation into an equivalent, more efficient form.
+
+  - Several relocations may result in an addition with zero, which may be optimized as follows:
+
+    ::
+
+      ADD x0, x1, 0   // eg. R_<CLS>_TLSLE_ ADD_TPREL_HI12
+      ADD x2, x2, 0   // or R_<CLS>_ADD_ABS_LO12_NC
+
+      // after optimization:
+
+      MOV x0, x1
+      NOP
+
+  - The relocation ``R_<CLS>_ADR_PREL_PG_HI21`` may emit a MOV with zero immediate for undefined weak symbols.
+
+- A sequence of relocations can be relaxed into more efficient relocations. All of the following conditions must be true:
+
+  - The relocations apply to consecutive instructions in the order specified.
+  - The relocations use the same symbol.
+  - The relocated instructions have the same source and destination register.
+  - The relocations do not appear separately or in a different order.
+
+  In this case each set of relocations is independent and may be optimized. The following sequences are defined:
+
+  - Large GOT indirection
+
+    ::
+
+      ADRP  x0, :got: symbol            // R_<CLS>_ADR_GOT_PAGE
+      LDR   x0, [x0 :got_lo12: symbol]  // R_<CLS>_LD64_GOT_LO12_NC
+
+      // after optimization:
+
+      ADRP  x0, symbol                  // R_<CLS>_ADR_PREL_PG_HI21
+      ADD   x0, x0 :lo12: symbol        // R_<CLS>_ADD_ABS_LO12_NC
+
+    A sequence may be optimized if it meets all of the following conditions:
+
+    - ``symbol`` is a non-preemtable definition.
```

Comment:
In the rest of the document we've used pre-empt rather than preempt. Would be good to keep that consistent. I guess non-pre-empable would work. Or we could reword to avoid the `non` with something like ``symbol`` is a definition that cannot be pre-empted.

Reply:
I'll fix